File: 3585.find-weighted-median-node-in-tree.go
#
# @lc app=leetcode id=3585 lang=golang
#
# [3585] Find Weighted Median Node in Tree
#

# @lc code=start
func findMedian(n int, edges [][]int, queries [][]int) []int {
	const LOG = 18
	type Pair struct {
		to int
		w  int64
	}
	adj := make([][]Pair, n)
	for _, e := range edges {
		u, v := e[0], e[1]
		w := int64(e[2])
		adj[u] = append(adj[u], Pair{v, w})
		adj[v] = append(adj[v], Pair{u, w})
	}
	parent := make([][]int, LOG)
	distPar := make([][]int64, LOG)
	depth := make([]int, n)
	distRoot := make([]int64, n)
	for i := range parent {
		parent[i] = make([]int, n)
		for j := range parent[i] {
			parent[i][j] = -1
		}
	}
	for i := range distPar {
		distPar[i] = make([]int64, n)
	}
	// BFS from root 0
	visited := make([]bool, n)
	q := make([]int, 0, n)
	qi := 0
	parent[0][0] = -1
	depth[0] = 0
	distRoot[0] = 0
	visited[0] = true
	q = append(q, 0)
	for qi < len(q) {
		u := q[qi]
		qi++
		for _, pr := range adj[u] {
			v := pr.to
			w := pr.w
			if !visited[v] {
				visited[v] = true
				parent[0][v] = u
				distPar[0][v] = w
				depth[v] = depth[u] + 1
				distRoot[v] = distRoot[u] + w
				q = append(q, v)
			}
		}
	}
	// Build lifting tables
	for k := 1; k < LOG; k++ {
		for i := 0; i < n; i++ {
			p := parent[k-1][i]
			if p != -1 {
				pp := parent[k-1][p]
				parent[k][i] = pp
				if pp != -1 {
					distPar[k][i] = distPar[k-1][i] + distPar[k-1][p]
				}
			}
		}
	}
	lcaFunc := func(u, v int) int {
		if depth[u] > depth[v] {
			u, v = v, u
		}
		// Equalize depths
		diff := depth[v] - depth[u]
		for k := 0; k < LOG; k++ {
			if (diff & (1 << k)) != 0 {
				v = parent[k][v]
			}
		}
		if u == v {
			return u
		}
		for k := LOG - 1; k >= 0; k-- {
			if parent[k][u] != parent[k][v] {
				u = parent[k][u]
				v = parent[k][v]
			}
		}
		return parent[0][u]
	}
	ans := make([]int, len(queries))
	for j := range queries {
		u := queries[j][0]
		v := queries[j][1]
		if u == v {
			ans[j] = u
			continue
		}
		l := lcaFunc(u, v)
		dul := distRoot[u] - distRoot[l]
		dlv := distRoot[v] - distRoot[l]
		total := dul + dlv
		target := (total + 1) / 2
		if dul >= target {
			// Lift from u
			cur := u
			cum := int64(0)
			needed := target
			for k := LOG - 1; k >= 0; k-- {
				if parent[k][cur] != -1 && cum + distPar[k][cur] < needed {
					cum += distPar[k][cur]
					cur = parent[k][cur]
				}
			}
			x := parent[0][cur]
			ans[j] = x
		} else {
			// Lift from v
			addNeeded := target - dul
			maxAllow := dlv - addNeeded
			cur := v
			cum := int64(0)
			for k := LOG - 1; k >= 0; k-- {
				if parent[k][cur] != -1 && cum + distPar[k][cur] <= maxAllow {
					cum += distPar[k][cur]
					cur = parent[k][cur]
				}
			}
			ans[j] = cur
		}
	}
	return ans
}
# @lc code=end

File: 3585.find-weighted-median-node-in-tree.java
#
# @lc app=leetcode id=3585 lang=java
#
# [3585] Find Weighted Median Node in Tree
#

# @lc code=start
class Solution {
    public int[] findMedian(int n, int[][] edges, int[][] queries) {
        java.util.List<int[]>[] adj = new java.util.ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new java.util.ArrayList<int[]>();
        }
        for (int[] e : edges) {
            int u = e[0], v = e[1], w = e[2];
            adj[u].add(new int[]{v, w});
            adj[v].add(new int[]{u, w});
        }
        final int LOG = 18;
        long[] depth = new long[n];
        int[] level = new int[n];
        int[][] par = new int[n][LOG];
        long[][] updist = new long[n][LOG];
        for (int i = 0; i < n; i++) {
            java.util.Arrays.fill(par[i], -1);
        }
        // BFS for depth, level, par[0], updist[0]
        boolean[] vis = new boolean[n];
        java.util.Queue<Integer> q = new java.util.LinkedList<>();
        vis[0] = true;
        depth[0] = 0;
        level[0] = 0;
        par[0][0] = -1;
        updist[0][0] = 0;
        q.offer(0);
        while (!q.isEmpty()) {
            int node = q.poll();
            for (int[] pr : adj[node]) {
                int v = pr[0];
                int w = pr[1];
                if (!vis[v]) {
                    vis[v] = true;
                    par[v][0] = node;
                    updist[v][0] = w;
                    depth[v] = depth[node] + w;
                    level[v] = level[node] + 1;
                    q.offer(v);
                }
            }
        }
        // Fill lifting tables
        for (int k = 1; k < LOG; k++) {
            for (int i = 0; i < n; i++) {
                int a = par[i][k - 1];
                if (a != -1) {
                    par[i][k] = par[a][k - 1];
                    updist[i][k] = updist[i][k - 1] + updist[a][k - 1];
                } else {
                    par[i][k] = -1;
                    updist[i][k] = 0;
                }
            }
        }
        int[] ans = new int[queries.length];
        for (int j = 0; j < queries.length; j++) {
            int u = queries[j][0];
            int v = queries[j][1];
            if (u == v) {
                ans[j] = u;
                continue;
            }
            // Compute LCA using levels
            int aa = u, bb = v;
            if (level[aa] > level[bb]) {
                int t = aa;
                aa = bb;
                bb = t;
            }
            // Lift bb to level[aa]
            for (int k = LOG - 1; k >= 0; k--) {
                if (par[bb][k] != -1 && level[par[bb][k]] >= level[aa]) {
                    bb = par[bb][k];
                }
            }
            int lca_;
            if (aa == bb) {
                lca_ = aa;
            } else {
                for (int k = LOG - 1; k >= 0; k--) {
                    if (par[aa][k] != -1 && par[bb][k] != -1 && par[aa][k] != par[bb][k]) {
                        aa = par[aa][k];
                        bb = par[bb][k];
                    }
                }
                lca_ = par[aa][0];
            }
            long total = depth[u] + depth[v] - 2L * depth[lca_];
            long targ = (total + 1L) / 2L;
            long distul = depth[u] - depth[lca_];
            int med;
            if (distul >= targ) {
                long maxd = targ - 1L;
                int z = u;
                long remd = maxd;
                for (int k = LOG - 1; k >= 0; k--) {
                    if (par[z][k] != -1 && updist[z][k] <= remd) {
                        remd -= updist[z][k];
                        z = par[z][k];
                    }
                }
                med = par[z][0];
            } else {
                long distlv = depth[v] - depth[lca_];
                long rem = targ - distul;
                long maxupv = distlv - rem;
                int y = v;
                long remd = maxupv;
                for (int k = LOG - 1; k >= 0; k--) {
                    if (par[y][k] != -1 && updist[y][k] <= remd) {
                        remd -= updist[y][k];
                        y = par[y][k];
                    }
                }
                med = y;
            }
            ans[j] = med;
        }
        return ans;
    }
}
# @lc code=end